cpp/src/deriv.cpp

```diff
@@ -26,3 +26,38 @@ LinearOperator M_operator(const SparseMatrix &Q, const std::vector<Cone> &cones,
   return (aslinearoperator(Q) - identity(N)) * dpi(u, v, w, cones) +
          identity(N);
 }
+
+Matrix dpi_dense(const Vector &u, const Vector &v, double w,
+                 const std::vector<Cone> &cones) {
+  int n = u.size();
+  int m = v.size();
+  int N = n + m + 1;
+  Matrix D = Matrix::Zero(N, N);
+  D.block(0, 0, n, n).diagonal().array() += 1.;
+  // Could be optimized by having dprojection_dense modifying this in-place,
+  // or by not explicitly adding the first and last blocks.
+  D.block(n, n, N - n - 1, N - n - 1) = dprojection_dense(v, cones, true);
+  D(N - 1, N - 1) = gt(w, 0.0);
+  return D;
+}
+
+Matrix M_dense(const Matrix &Q, const std::vector<Cone> &cones, const Vector &u,
+               const Vector &v, double w) {
+  int n = u.size();
+  int m = v.size();
+  int N = n + m + 1;
+  Matrix eye = Matrix::Identity(N, N);
+  return (Q - eye) * dpi_dense(u, v, w, cones) + eye;
+}
+
+Vector _solve_derivative_dense(const Matrix &M, const Matrix &MT,
+                               const Vector &rhs) {
+  // TODO: Factorization could be cached to optimize multiple calls
+  return (MT * M).ldlt().solve(MT * rhs);
```

Comment:
Can we do QR here? MT*M might be singular. I think ldlt() silently fails. I'd be OK with ldlt if we have a backup plan.

Reply:
Another benefit of QR is we don't need M and MT.

Reply:
In some quick tests QR seemed slightly slower. What do you think about adding a small ridge here, or detecting the conditioning of M MT and falling back to QR if it's below some threshold?

Reply:
LDLT is fine. @sbarratt says it's fine as well.

Reply:
Yeah LDLT should actually always work.